### ISSUE TITLE
JSR203 NIO FileSystemProvider?

### ISSUE THREAD

**Issue by ryan-williams** [Azure/azure-storage-java#305]
Has one been implemented for Azure storage? Just wondering, thanks!

**Comment by rickle-msft** [comment]
Hey, Ryan!

 We aren't adhering to NIO2. Instead we're following Reactive Streams through RxJava. We actually just released a preview of the new version of the SDK which offers this. You can find it here: https://github.com/Azure/azure-storage-java/tree/New-Storage-SDK-V10-Preview

**Comment by ryan-williams** [comment]
Good to know, thanks… that's not exclusive with supporting an NIO interface though, right? In fact, presumably a thin NIO facade over what's already there would be straightforward.

Can reinterpret this issue as a request for that 😄

**Comment by rickle-msft** [comment]
Possibly not. I think this will have to be a discussion for our team sometime after GA of the new version. I'll add an item to the backlog to ensure a discussion on the best way to support this scenario.

**Comment by pditommaso** [comment]
+1 for having JSR203 NIO FileSystemProvider for Azure storage. That would allow to plug Azure cloud storage in Java applications as a plain file system.

**Comment by fstrozzi** [comment]
A JSR203 NIO FileSystemProvider will be extremely useful, and would allow to use the Azure storage alongside local storage in a transparent way. We rely on NIO interface also on other cloud storage backends and the advantages in terms of usability are great.

**Comment by stewartadam** [comment]
This would be useful for genomics scenarios as well, where some JVM-based workflow engines abstract cloud vs local file access using NIO.

**Comment by pditommaso** [comment]
> This would be useful for genomics scenarios as well, where some JVM-based workflow engines abstract cloud vs local file access using NIO.

I think all people in this thread are exactly referring to this scenario.

**Comment by rickle-msft** [comment]
Hi, guys. 

Quick update. Since this issue has continuously been referenced by customers, I want to work on at least prioritizing the discussion surrounding this to work towards a more conclusive answer on our path forward.

We GA'd the library a couple months ago, which was the first benchmark we wanted to hit before looking closer at this. There are a few features we didn't get into GA that we still feel are rather critical that we want to be sure to complete soon. Many of these have been completed by now, but those (and a few bug reports we've prioritized) have been the bulk of development since August. The most notable remaining critical item is uploadFromNonReplayableFlowable, which I should get to start working on next week or possibly the week after. Once that is finished (ideally speaking, that would be maybe a week or two of work), there would be potentially more room for discussion, depending on when the service ends up coming out with new features and how critical those are to support. So that's sort of the landscape of dev time.

I noticed @sodic mentioned that he was taking a look at implementing this himself. Would he or anyone else here be willing to discuss the possibility of a pull request? We should discuss design and management and whether this repo is the correct place for such a feature before committing to anything, but if you're willing to contribute, we might be able to work something out. 

Would anyone interested in this feature be willing to hop on a call to discuss the details of the scenario and the work involved in supporting this story? I'm not super familiar with JSR203 and it would certainly help jumpstart the conversation if I could get a primer.

In the meanitime, is it strongly required that JSR203 be used in these scenarios? Just brainstorming here, but Azure Storage does have a few other options that work with filesystems in this sort of transparent manner. Fuse and SMB are solutions for mounting a Linux or Windows filesystem respectively. Without having more details, it sounds like those might be possible options. They don't interact with the Java client in any special way, but they both transparently back a file system on azure storage, so any file you write or operate on locally is reflected as a blob in Storage.

Thanks, guys, for your continued interest.

**Comment by pditommaso** [comment]
Hi Rick, thanks for your detailed reply. I'll provide my feedback. 

Regarding the possibility to use alternative adapters, such Fuse, that's something I've investigated in the past, but none was satisfying. For example Fuse requires the installation of separate OS libraries, which need root permissions, and the compilation of the fuse client. Operations that are overkill for average users. Also the client requires separate configuration and does not integrated smoothly with the target application. 

I would surely be interested in this effort, I've implemented a couple of JSR203 file system drivers. It's not a particular friendly API but neither so complicated. The most challenging part is to have a detailed knowledge of the target API and storage system ie. Azure blob storage, that I guess  you have. 

If you are interested to collaborate on this I would happy to provide some advice and to contribute my implementation for [Google storage](https://github.com/seqeralabs/nxf-gsfs). It can be used as a reference implementation for the Azure version (it's groovy code, but converting to Java it would be trivial). Let me know what do you think about that.

**Comment by rickle-msft** [comment]
@pditommaso. Thanks for the feedback! I've blocked off some time next sprint (starting next week) to do some more investigation and possibly design. Since you're interested in this effort, is there a channel where we could set up a call or an IM chat to discuss some details? Perhaps mid-week would be best so I can do some reading beforehand.

**Comment by pditommaso** [comment]
Nice. I've created [this gitter](https://gitter.im/Azure-jsr203/Lobby#) channel. We may organise also a call something Wed or Thu.

**Comment by stewartadam** [comment]
@rickle-msft was there any outcomes from the discussion?

**Comment by rickle-msft** [comment]
Yes, but nothing final. This feature is seen by our management as sort of a convenience thing, so it's been prioritized lower than work needed for supporting service features and GAing some other tools. Our goal is to post a sort of design doc here once we have one that people can comment on, but I don't have a very solid timeline for that at the moment, unfortunately.

**Comment by saurabhchawla** [comment]
+1 for having JSR203 NIO FileSystemProvider for Azure storage.

**Comment by nalinigans** [comment]
This would be very useful to have, any updates?

**Comment by rickle-msft** [comment]
We got up to the point of pulling the initial work item into our sprint and then ended up having to defer it becuase several crucial stability issues arose in v11 at the same time and eventually led to this effort to rework things for v12 (now in preview). I completely recognize that this is probably our most requested feature for this sdk, so once we our confident in the stability of our baseline sdk, this will be a high priority for us (which I realize I've been saying for a year, but I promise I'm not just stalling).

**Comment by landro** [comment]
Any updates on this @rickle-msft ? Our customer needs this urgently

**Comment by rickle-msft** [comment]
Unfortunately not. Since the last update, we have been working on GAing v12, which is now out. Our management is currently doing some planning for the next semester starting in the new year, and I submitted this request to be considered for inclusion in that work. I don't have any word yet on whether or not it was accepted.

**Comment by landro** [comment]
Thanks for getting back to me, @rickle-msft !  I appreciate it.
BTW, here is some relevant info from [AWS](https://github.com/aws/aws-sdk-java-v2/issues/1388) and [GCS](https://github.com/googleapis/google-cloud-java/tree/master/google-cloud-clients/google-cloud-contrib/google-cloud-nio) that you might find relevant.

**Comment by rickle-msft** [comment]
Thank you for sharing that! I am sure it will speed up our development to have these references when we get started on this project. I know this has been eagerly anticipated for a while, so I am looking forward to getting started on it whenever I get a green light

**Comment by rickle-msft** [comment]
To all those still watching this thread, I will begin investigating this again sometime next week and will start on an implementation thereafter. I will continue to post updates of progress here.
